### PR TITLE
bpo-32399: Starting with AIX6.1 there is support in libc.a for uuid (RFC4122)

### PR DESCRIPTION
**Pull request by aixtools**
This patch provides the changes needed for this integration with the OS.

On AIX the base function is uuid_create() rather than uuid_generate_time()
The AIX uuid_t typedef is more aligned to the UUID field based definition
while the Linux typedef that is more aligned with UUID bytes
(or perhaps UUID bytes_le) definitions.

https://bugs.python.org/issue32399

### CODE REVIEW

File: Modules/_uuidmodule.c [hunk 1]
```diff
@@ -1,22 +1,30 @@
...
+    uuid_t uuid;
 #ifdef HAVE_UUID_GENERATE_TIME_SAFE
-    uuid_t out;
     int res;
 
-    res = uuid_generate_time_safe(out);
-    return Py_BuildValue("y#i", (const char *) out, sizeof(out), res);
+    res = uuid_generate_time_safe(uuid);
+    return Py_BuildValue("y#i", (const char *) uuid, sizeof(uuid), res);
+#elif HAVE_UUID_CREATE
```
Comment (pitrou) [inline Modules/_uuidmodule.c:21]: Could you add a comment that this is AIX-specific?

File: configure.ac [hunk 2]
```diff
@@ -2692,6 +2695,17 @@ void *x = uuid_generate_time_safe
   [AC_MSG_RESULT(no)]
 )
 
+AC_MSG_CHECKING(for uuid_create)
```
Comment (pitrou) [inline configure.ac]: Could you add a comment that this is AIX-specific?

File: Modules/_uuidmodule.c [hunk 3]
```diff
@@ -1,22 +1,30 @@
...
+    unsigned32 status;
+    uuid_create(&uuid, &status);
```
Comment (pitrou) [inline Modules/_uuidmodule.c]: You should check the return value of `uuid_create`, no?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the PR. A couple comments below.

**Comment by aixtools** [comment]
Regarding the return value checking. 
I could check that status != 0, but what should be done then. 

Could the argument ‘y#i’ and status included as an argument be used? I admit that I didn’t look for documentation on PyBuildValue()

**Comment by pitrou** [comment]
If there is an error, you should probably return `PyErr_SetFromErrno(PyExc_OSError)`. You can find examples of this in other modules in the `Modules` directory.

**Comment by aixtools** [comment]
I have made the requested changes; please review again.

File: configure [hunk 4]
```diff
@@ -9125,7 +9125,7 @@ then
 			LDCXXSHARED='$(CXX) -shared'
 		else
 			LDSHARED='$(CC) -b'
-			LDCXXSHARED='$(CXX) -shared'
+			LDCXXSHARED='$(CXX) -b'
```
Comment (pitrou) [inline configure:9128]: This doesn't look related.

File: configure.ac [hunk 5]
```diff
@@ -2681,6 +2681,10 @@ AC_CHECK_LIB(sendfile, sendfile)
 AC_CHECK_LIB(dl, dlopen)	# Dynamic linking for SunOS/Solaris and SYSV
 AC_CHECK_LIB(dld, shl_load)	# Dynamic linking for HP-UX
 
+# checks for uuid.h location
+dnl # AIX is at /usr/include/uuid.h
```
Comment (pitrou) [inline configure.ac]: "dnl" isn't useful here, no?

File: configure.ac [hunk 6]
```diff
@@ -2692,6 +2696,18 @@ void *x = uuid_generate_time_safe
   [AC_MSG_RESULT(no)]
 )
 
+# AIX provides support for RFC4122 (uuid) in libc.a starring with AIX 6.1 (anno 2007)
```
Comment (pitrou) [inline configure.ac]: Typo: "starting".

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you. Just a few more comments.

**Comment by aixtools** [comment]
On 30-Dec-17 13:21, Antoine Pitrou wrote:
>
> *@pitrou* requested changes on this pull request.
>
> Thank you. Just a few more comments.
>
> ------------------------------------------------------------------------
>
> In configure 
> <https://github.com/python/cpython/pull/4974#discussion_r159121770>:
>
> > @@ -9125,7 +9125,7 @@ then
>   			LDCXXSHARED='$(CXX) -shared'
>   		else
>   			LDSHARED='$(CC) -b'
> -			LDCXXSHARED='$(CXX) -shared'
> +			LDCXXSHARED='$(CXX) -b'
This is, I think, a side effect of the different aclocal.m4 that my 
autoreconf generates.  I'll have to read the docs, I think they are 
equivalent. But I'll check.
>
> This doesn't look related.
>
> ------------------------------------------------------------------------
>
> In configure.ac 
> <https://github.com/python/cpython/pull/4974#discussion_r159121779>:
>
> > @@ -2681,6 +2681,10 @@ AC_CHECK_LIB(sendfile, sendfile)
>   AC_CHECK_LIB(dl, dlopen)	# Dynamic linking for SunOS/Solaris and SYSV
>   AC_CHECK_LIB(dld, shl_load)	# Dynamic linking for HP-UX
>   
> +# checks for uuid.h location
> +dnl # AIX is at /usr/include/uuid.h
>
> "dnl" isn't useful here, no?
>
I was think the comment about AIX was relevant for the source, less so 
for the generated file.
I'll delete it as I reference AIX elsewhere.
> ------------------------------------------------------------------------
>
> In configure.ac 
> <https://github.com/python/cpython/pull/4974#discussion_r159121783>:
>
> > @@ -2692,6 +2696,18 @@ void *x = uuid_generate_time_safe
>     [AC_MSG_RESULT(no)]
>   )
>   
> +# AIX provides support for RFC4122 (uuid) in libc.a starring with AIX 6.1 (anno 2007)
>
> Typo: "starting".
>
fixed.

**Comment by pitrou** [comment]
> I was think the comment about AIX was relevant for the source, less so for the generated file.

I was tallking about the "dnl" at the beginning of the comment. I don't know much about m4, but it doesn't seem necessary, given how other comments are formatted.

File: configure (hunk 4 above)
Comment (pitrou) [reply configure:9128]: Note those lines originate from #2519. Perhaps @datalogics-robb can comment on why `LDCXXSHARED='$(CXX) -b'` in `configure.ac` becomes `LDCXXSHARED='$(CXX) -shared'` in `configure`.

**Comment by aixtools** [comment]
On 30-Dec-17 13:21, Antoine Pitrou wrote:
>
> *@pitrou* requested changes on this pull request.
>
> Thank you. Just a few more comments.
>
> ------------------------------------------------------------------------
>
> In configure 
> <https://github.com/python/cpython/pull/4974#discussion_r159121770>:
>
> > @@ -9125,7 +9125,7 @@ then
>   			LDCXXSHARED='$(CXX) -shared'
>   		else
>   			LDSHARED='$(CC) -b'
> -			LDCXXSHARED='$(CXX) -shared'
> +			LDCXXSHARED='$(CXX) -b'
>
> This doesn't look related.
>
imho - there is a commit error for configure, as configure.ac says:

         hp*|HP*)
                 if test "$GCC" = "yes" ; then
                         LDSHARED='$(CC) -shared'
                         LDCXXSHARED='$(CXX) -shared'
                 else
                         LDSHARED='$(CC) -b'
                         LDCXXSHARED='$(CXX) -b'
                 fi ;;
         Darwin/1.3*)
                 LDSHARED='$(CC) -bundle'

so my output is, as far as I can determine, correct.

**Comment by aixtools** [comment]
I have made the requested changes; please review again

**Review by pitrou — APPROVED** [review]
Ok, this looks good to me. I'm gonna merge once the CI is happy.

File: configure (hunk 4 above)
Comment (aixtools) [reply configure:9128]: I looked at #2519 and it looks like the changes in configure were editted manually, rather than generated from configure.ac (my two cents).

If the line is suppossed to be with -shared, then configure.ac needs to be fixed. I do not know HPUX, so cannot help.

**Comment by pitrou** [comment]
Thank you @aixtools !

File: configure (hunk 4 above)
Comment (datalogics-robb) [reply configure:9128]: Sorry I've been AFK for a week, but yes this change looks good.  -shared is a gcc/g++ flag and should definitely not be in a non-gcc clause.
